### PR TITLE
Spark: Quote only required attributes

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR optimizes the quoting in Spark and addresses [this](https://github.com/apache/iceberg/pull/3529#discussion_r746876107) comment.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
cc @flyrain @rdblue @RussellSpitzer @karuppayya @szehon-ho @kbendick

**Comment by RussellSpitzer** [comment]
I'm not sure the added complexity here is worth the savings. It just feels like even with thousands of columns this should be on the order of milliseconds?

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkSchemaUtil.java [hunk 1]
```diff
@@ -319,7 +319,12 @@ public static void validateMetadataColumnReferences(Schema tableSchema, Schema r
   }
 
   public static Map<Integer, String> indexQuotedNameById(Schema schema) {
+    return indexQuotedNameById(schema, null);
+  }
+
+  public static Map<Integer, String> indexQuotedNameById(Schema schema, Set<Integer> fieldIds) {
     Function<String, String> quotingFunc = name -> String.format("`%s`", name.replace("`", "``"));
-    return TypeUtil.indexQuotedNameById(schema.asStruct(), quotingFunc);
+    Schema projectedSchema = fieldIds != null ? TypeUtil.select(schema, fieldIds) : schema;
```
[the remark this answers is not in the document]
Comment (aokolnychyi) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkSchemaUtil.java:327]: It depends. I think the biggest performance hit is calling replace on the underlying strings. We may pass a set field ids to the indexing util or just adapt the quoting function to use a regex to see if the column should be quoted.

[the remark this answers is not in the document]
Comment (aokolnychyi) [reply spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkSchemaUtil.java:327]: It depends. I think the biggest performance hit is calling replace on the underlying strings. We may pass a set field ids to the indexing util or just adapt the quoting function to use a regex to see if the column should be quoted.

**Comment by aokolnychyi** [comment]
Yeah, I am not convinced either. I did not have it in the first place but I thought I would trigger this discussion.

File: spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkSchemaUtil.java [hunk 2]
```diff
@@ -319,7 +319,12 @@ public static void validateMetadataColumnReferences(Schema tableSchema, Schema r
   }
 
   public static Map<Integer, String> indexQuotedNameById(Schema schema) {
+    return indexQuotedNameById(schema, null);
+  }
+
+  public static Map<Integer, String> indexQuotedNameById(Schema schema, Set<Integer> fieldIds) {
```
Comment (rdblue) [inline spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/SparkSchemaUtil.java:325]: Do we need the original method if we are passing `fieldIds` in all internal uses? It hasn't been in a release so we can just remove it.

**Comment by rdblue** [comment]
I agree that string manipulation is probably more expensive than the select, but I'm not too concerned about this.

That said, the complexity here looks reasonable since it is just passing a set of IDs to `select`. I could go either way. +1 if you want to merge it.

**Comment by aokolnychyi** [comment]
Okay, I did a quick benchmark and the new approach does give us a 8x improvement but the current call even on a schema with 1k columns completes in a millisecond. Taking this into account, I don't think it is reasonable to complicate things. I am closing this PR.

**Pull request closed without merging**